### PR TITLE
Fix cross-version-tests for Keras and Tensorflow

### PR DESCRIPTION
**Pull request by daniellok-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/daniellok-db/mlflow/pull/11001?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11001/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11001
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

This PR fixes a few CI issues:
1. [tensorflow cross-version tests are failing](https://github.com/mlflow-automation/mlflow/actions/workflows/cross-version-tests.yml)
  - this is because the tensorflow tests save keras models with keras 2, and we recently updated the `mlflow.keras` model to support keras 3
2. [keras-dev tests are failing](https://github.com/mlflow-automation/mlflow/actions/runs/7774344898/job/21201727613#step:15:63)
  - it seems that the dev version of keras has a dependency on tensorflow

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: .github/workflows/cross-version-tests.yml [hunk 1]
```diff
@@ -185,6 +185,10 @@ jobs:
           pip install -U "pip$PIP_SPEC" wheel
           pip install -e .[extras]
           pip install -r requirements/test-requirements.txt
+      - name: Install tensorflow for keras
```
Comment (daniellok-db) [inline .github/workflows/cross-version-tests.yml:188]: looks like keras dev has some dependency on tensorflow:

https://github.com/mlflow/mlflow/actions/runs/7778504032/job/21208301502#step:15:58

File: mlflow/keras/__init__.py [hunk 2]
```diff
@@ -1,24 +1,45 @@
...
-from mlflow.keras.callback import MLflowCallback
-from mlflow.keras.load import _load_pyfunc, load_model
-from mlflow.keras.save import (
-    get_default_conda_env,
-    get_default_pip_requirements,
-    log_model,
-    save_model,
-)
+if Version(keras.__version__) < Version("3.0.0"):
+    from mlflow.tensorflow import (
```
Comment (daniellok-db) [inline mlflow/keras/__init__.py:6]: undo the deletion from the original PR, keeping it behind a version check instead:

https://github.com/mlflow/mlflow/pull/10830/files#diff-6e6f5be585845fe6e5d51023750107215d9d70d0f5039ad0a5c8a4b570c741d8L1-L9

File: requirements/gateway-requirements.txt [hunk 3]
```diff
@@ -3,7 +3,7 @@
 # dev/generate_requirements.py script to regenerate this file.
 
 pydantic<3,>=1.0
-fastapi<1
+fastapi<=0.109.2
```
Comment (daniellok-db) [inline requirements/gateway-requirements.txt:6]: unrelated to keras, but the `deployments` CI job is also failing due to `fastapi` upgrading their version of `starlette` to 0.36.3: https://github.com/tiangolo/fastapi/pull/11086

`starlette>=0.36.0` now [throws when passed an invalid config file](https://github.com/encode/starlette/pull/2422), and `slowapi` dependency by [default passes](https://github.com/laurentS/slowapi/blob/master/slowapi/extension.py#L158) in `".env"` as a file if none is provided.

File: requirements/gateway-requirements.txt [hunk 4]
```diff
@@ -3,7 +3,13 @@
...
 
 pydantic<3,>=1.0
-fastapi<1
+# fastapi 0.109.2 upgraded their `starlette` dependency,
+# which has a breaking interaction with `slowapi` when
+# the ".env" file doesn't exist.
+# https://github.com/laurentS/slowapi/pull/184 has been
+# filed to address this, but until it is approved we need
+# to pin the fastapi dependency
+fastapi<0.109.1
```
Comment (daniellok-db) [inline requirements/gateway-requirements.txt:12]: i saw i'm supposed to edit `dev/generate_requirements` to update this, but it doesn't allow us to specify a max patch version. i can use `unsupported` in the YAML file but then we have to update this with every single new fastapi release.

Comment (daniellok-db) [reply requirements/gateway-requirements.txt:12]: oh nevermind there is CI forcing us to edit the YAML file

**Comment by minkj1992** [comment]
@daniellok-db  It seems that the work you're doing is related to mine. Since my PR may not get approved, could you possibly add the keras nitpick to this PR?

https://github.com/mlflow/mlflow/pull/11003

**Comment by daniellok-db** [comment]
> @daniellok-db It seems that the work you're doing is related to mine. Since my PR may not get approved, could you add the keras nitpick_ignore to this PR?
> 
> #11003

@minkj1992 ah perfect! Sorry I missed that PR, yes I have to update the keras RST file for the docs anyway so i'll make sure to include this. thank you!

File: .github/workflows/cross-version-tests.yml (hunk 1 above)
Comment (WeichenXu123) [reply .github/workflows/cross-version-tests.yml:188]: Issue:

If you install tensorflow, it will overwrite existing keras version to 2.15
see https://keras.io/getting_started/

Comment (daniellok-db) [reply .github/workflows/cross-version-tests.yml:188]: i think this should install it before, since it's before the `- name: Install ${{ matrix.package }} ${{ matrix.version }}` line

File: mlflow/keras/__init__.py (hunk 2 above)
Comment (daniellok-db) [reply mlflow/keras/__init__.py:6]: cc @chenmoneygithub does this look alright to you?

File: .github/workflows/cross-version-tests.yml (hunk 1 above)
Comment (WeichenXu123) [reply .github/workflows/cross-version-tests.yml:188]: Oh you are right ^^^, thanks :)

**Review by serena-ruan — APPROVED** [review]
LGTM though I would prefer splitting different issues into different PRs 🤔

**Comment by daniellok-db** [comment]
> LGTM though I would prefer splitting different issues into different PRs 🤔

i can split the fastapi to a separate one!

File: mlflow/keras/__init__.py (hunk 2 above)
Comment (daniellok-db) [reply mlflow/keras/__init__.py:6]: sounds good! is it cool if i merge this first to unblock x-version tests? once you get around to making the difference in the transformers flavor, you can delete this here.

File: .github/workflows/cross-version-tests.yml (hunk 1 above)
Comment (chenmoneygithub) [reply .github/workflows/cross-version-tests.yml:188]: The dev issue should be resolved by pointing keras-nightly to the github head. 

```
  package_info:
    pip_release: "keras"
    install_dev: |
      pip install --upgrade git+https://github.com/keras-team/keras.git
```

For some reason keras-nightly is pointing to keras 2.15

**Review by chenmoneygithub — APPROVED** [review]
Thanks Daniel! One comment on the package installation

**Comment by harupy** [comment]
@daniellok-db I re-ran cross version tests with the `enable-dev-tests` label. It looks like we need to use python 3.9 for keras dev.
